### ISSUE TITLE
Magento v2.3 - Choosing a child theme for product .. layout xml doesn't work nor does the layout update

### ISSUE THREAD

**Issue by PaulWatt** [magento/magento2#26295]
### Preconditions (*)

1. Magento 2.3.x & 2.4-develop
2. PHP 7.2

### Steps to reproduce (*)
When you apply *ThemeC* on a product, then only the (extended) layout of *ThemeB* will be taken in the following scenario:

```
VendorA
  |__/ThemeA (parent: Magento/blank)
    |__/Magento_Catalog
      |__/layout
        |--catalog_product_view.xml

VendorB
  |__/ThemeB (parent: VendorA/ThemeA; applied theme for store)
    |__/Magento_Catalog
      |__/layout
        |--catalog_product_view.xml
  |__/ThemeC (parent: VendorB/ThemeB)
    |__/Magento_Catalog
      |__/layout
        |--catalog_product_view.xml
```

I don't know if there is any difference when all themes are by the same vendor, but at least this is my setup.
Even tough above setup describes extending layouts I've also tried to override the layout with the same effect.

**Expected result**
Magento pulls the "catalog_product_view.xml" from the child theme(theme C).

**Actual result**
Magento pulls the "catalog_product_view.xml" from the theme B.

-----------------------------------------------
**Additional information**
Currently I have fixed the issue with this [patch provided by kirmorozov](https://github.com/magento/magento2/issues/8333#issuecomment-363814837).

**Comment by engcom-Hotel** [comment]
Hello @PaulWatt 
```
Child_theme
-- theme
--- Magento_Catalog
---- layout
----- catalog_product_view.xml
```
structure is not valid for overriding theme layout.
How to override theme layouts you can read in this [topic](https://devdocs.magento.com/guides/v2.3/frontend-dev-guide/layouts/layout-override.html).

So i have to close this issue.

Thanks for your report and collaboration!

**Comment by felly** [comment]
Hello @engcom-Hotel

I have to disagree with you and in my opinion this is still a valid issue - reproduced on my side on 2.2.x and 2.3.x

When you apply *ThemeC* on a product, then only the (extended) layout of *ThemeB* will be taken in the following scenario:

```
VendorA
  |__/ThemeA (parent: Magento/blank)
    |__/Magento_Catalog
      |__/layout
        |--catalog_product_view.xml

VendorB
  |__/ThemeB (parent: VendorA/ThemeA; applied theme for store)
    |__/Magento_Catalog
      |__/layout
        |--catalog_product_view.xml
  |__/ThemeC (parent: VendorB/ThemeB)
    |__/Magento_Catalog
      |__/layout
        |--catalog_product_view.xml
```

I don't know if there is any difference when all themes are by the same vendor, but at least this is my setup.
Even tough above setup describes extending layouts I've also tried to override the layout with the same effect.

Currently I have fixed the issue with this [patch provided by kirmorozov](https://github.com/magento/magento2/issues/8333#issuecomment-363814837).

**Comment by engcom-Hotel** [comment]
Hello @felly , thanks for contributing!
I was able to reproduce this issue on the latest Magento 2.4-develop with your theme configuration.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Hotel 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-31174` were created

**Issue Available**: @engcom-Hotel, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by leonhelmus** [comment]
@felly Thank you very much for reporting this issue to me! 

I was going crazy why i couldn't see my changes of my child theme when changing the layout. 

This is still and issue in Magento 2.3.4.

**Comment by omaxmo** [comment]
@felly @PaulWatt @leonhelmus  Do you use the default theme(Applied Theme) for store view? This is a required step if you want a theme to be used on a storefront.

https://devdocs.magento.com/guides/v2.4/frontend-dev-guide/themes/theme-apply.html

Please set ThemeC as default for store view. This is should fix the issue.

**Comment by felly** [comment]
Hi @omaxmo 
the default/applied theme is (and should be) ThemeB - however the theme on some products should be ThemeC.

https://docs.magento.com/user-guide/catalog/settings-advanced-design.html
https://docs.magento.com/user-guide/catalog/product-scheduled-design-update.html

**Comment by Prunecreation** [comment]
Hi,
i had the same problem, (changing theme design in category or product doesn't apply the new theme layout for the product page).
I succeed in making this work using a before plugin on the execute function of  Catalog\Product\View controller .
Here is my code:
`<?php

namespace Vendor\Module\Plugin\Catalog\Product;

use \Magento\Framework\App\Action\Context;
/**
 * Disable customer email depending on settings value.
 *
 * @SuppressWarnings(PHPMD.UnusedFormalParameter)
 */
class View
{
    /**
     * @var \Magento\Framework\App\RequestInterface
     */
    protected $_request;

    protected $_objectManager;
    /**
     * @var Magento\Catalog\Model\Design
     */
    private $catalogDesign;
    /**
     * @var \Magento\Catalog\Api\ProductRepositoryInterface
     */
    private $productRepository;


    public function __construct(
        Context $context,
        \Magento\Catalog\Model\Design $catalogDesign,
        \Magento\Catalog\Api\ProductRepositoryInterface $productRepository
    ) {
        $this->_request = $context->getRequest();
        $this->_objectManager = $context->getObjectManager();
        $this->catalogDesign = $catalogDesign;
        $this->productRepository = $productRepository;
    }
    /**
     * Custom fix for https://github.com/magento/magento2/issues/7710
     * It allows to use a custom them for product page
     * @param View $subject
     * @see \Magento\Catalog\Controller\Product\View::execute
     */
    public function beforeExecute(\Magento\Catalog\Controller\Product\View $subject)
    {


        try {
            $productId = (int)$this->getRequest()->getParam('id');
            $product = $this->productRepository->getById($productId);

            if ($product) {
                $settings = $this->catalogDesign->getDesignSettings($product);

                if ($settings->getCustomDesign()) {
                    $this->catalogDesign->applyCustomDesign($settings->getCustomDesign());
                }
            }
        }
        catch(Exception $e){

        }
    }

    /**
     * Retrieve request object
     *
     * @return \Magento\Framework\App\RequestInterface
     */
    public function getRequest()
    {
        return $this->_request;
    }
}`

Don't forget to add the plugin definition in the di.xml of your module.

Let me know if it works for you.

Prunecreation

**Comment by leonhelmus** [comment]
I'm curious will this be fixed in 2.4.1?

**Comment by antoniocarboni** [comment]
The bug is stil existing on 2.4.2 :(

**Comment by extroniks** [comment]
This is still an issue for me also.

Here is a workaround I applied on our project as it was blocking development, it is a not so nice way of addressing the issue as I didn't have time to allocate more time into resolving the issue and it was easier to just bypass the issue.

Hopefully it helps someone here until a proper fix is made:
https://github.com/extroniks/module-catalog-theme-inherit-fix

**Comment by ihor-sviziev** [comment]
Hi @PaulWatt,
According to https://github.com/magento/magento2/pull/33825#issuecomment-901738129, it looks like internal team already fixed this issue in the 2.4-develop branch, related commits:
- https://github.com/magento/magento2/commit/7f6e9a842c3ca0d1971920c44d2e84f51b8bb5f8
- https://github.com/magento/magento2/commit/1391203bb31ce6ee1fba91aa9c89ceecddd5c621
- https://github.com/magento/magento2/commit/76825d1b61cd18ac91683f615c31de163a472f89

Could you confirm that?

**Comment by engcom-Bravo** [comment]
@ihor-sviziev Thanks for your contribution!!.

As per this comment https://github.com/magento/magento2/issues/26295#issuecomment-901786141.since the issue was fixed we are closing this issue.

Thanks.
